### PR TITLE
D/jbartolotta/w 7452168 show error

### PR DESCRIPTION
**Pull request by jbartolotta-sfdc**
This PR has a new branch name and contains the same content as https://github.com/forcedotcom/sfdx-scanner/pull/106

### CODE REVIEW

File: pmd-cataloger/src/main/java/sfdc/sfdx/scanner/xml/XmlReader.java [hunk 1]
```diff
@@ -26,10 +30,22 @@ public static XmlReader getInstance() {
 	 *
 	 * @param path - The path to a resource.
 	 * @return - An InputStream for the provided resource.
+	 * @throws FileNotFoundException - if path is a regular file but that file can't be read
 	 */
-	private InputStream getResourceAsStream(String path) {
-		final InputStream in = Thread.currentThread().getContextClassLoader().getResourceAsStream(path);
-		return in == null ? getClass().getResourceAsStream(path) : in;
+	private InputStream getResourceAsStream(String path) throws FileNotFoundException {
+		InputStream in = Thread.currentThread().getContextClassLoader().getResourceAsStream(path);
```
Comment (jfeingold35) [inline pmd-cataloger/src/main/java/sfdc/sfdx/scanner/xml/XmlReader.java:36]: As someone who's not terribly familiar with IO in Java, I'd love to see some comments explaining what's going on here.

File: pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/XmlFileFinder.java [hunk 2]
```diff
@@ -34,39 +35,66 @@
 		}
 	}
 
+	/**
+	 * Represents a container for XML. This can either be a Jar file which may contain multiple XML files or an XML
+	 * file that contains itself.
+	 */
+	public static final class XmlContainer {
+		public final String filePath;
+		public final List<String> containedFilePaths;
```
Comment (jfeingold35) [inline pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/XmlFileFinder.java:44]: I'd love to see commenting for what these variables mean. Like `filePath is the path to the file, and if the file is a JAR then containedFilePaths will be all the XMLs inside that JAR`, or something like that.

File: messages/EventKeyTemplates.js [hunk 3]
```diff
@@ -26,7 +26,8 @@ module.exports = {
 			"multipleRuleDesc": "ERROR: PMD Rule [%s] has %s 'description' elements. Please reduce this number to 1.",
 			"recursionLimitReached": "ERROR: PMD Ruleset [%s] references rule [%s] through 10 or more layers of indirection. Please reduce this number.",
 			"xmlNotReadable": "ERROR: Error occurred while reading file [%s]: %s",
-			"xmlNotParsable": "ERROR: Could not parse XML file [%s]: %s"
+			"xmlNotParsable": "ERROR: Could not parse XML file [%s]: %s",
+			"duplicateXmlPath": "XML path [%s] defined in jar [%s] collides with previously defined path in jar [%s]. You will need to remove one of the jars by executing the following command 'sfdx scanner:rule:remove --force --path <jar-to-remove>'"
```
[the remark this answers is not in the document]
Comment (jbartolotta-sfdc) [reply messages/EventKeyTemplates.js:30]: Thanks! will fix.

File: pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/PmdRuleCataloger.java [hunk 4]
```diff
@@ -22,7 +22,7 @@
 	private Map<String, List<String>> rulePathEntries;
 
 	// Holds category and rulesets maps that provide files we need to scan for each language.
-	private LanguageXmlFileMapping languageXmlFileMapping = LanguageXmlFileMapping.getInstance();
+	private LanguageXmlFileMapping languageXmlFileMapping = new LanguageXmlFileMapping();
```
[the remark this answers is not in the document]
Comment (jbartolotta-sfdc) [reply pmd-cataloger/src/main/java/sfdc/sfdx/scanner/pmd/PmdRuleCataloger.java:25]: Sorry, PR https://github.com/forcedotcom/sfdx-scanner/pull/106 has more annotations with overall info on the changes. I didn't copy them over here. 

A singleton was problematic for unit tests, could add a #reset method, but then tests couldn't be run in parallel either.

File: pmd-cataloger/src/main/java/sfdc/sfdx/scanner/xml/XmlReader.java (hunk 1 above)
Comment (jbartolotta-sfdc) [reply pmd-cataloger/src/main/java/sfdc/sfdx/scanner/xml/XmlReader.java:36]: Good question. Let me see if I can find out why the code is looking at two class loaders.

File: src/lib/CustomRulePathManager.ts [hunk 5]
```diff
@@ -246,7 +246,7 @@ export class CustomRulePathManager implements RulePathManager {
 				throw SfdxError.create('@salesforce/sfdx-scanner', 'add', 'errors.invalidFilePath', [p]);
 			}
 			if (stats.isFile()) {
-				if (p.endsWith(".jar")) {
+				if (p.endsWith(".jar") || p.endsWith(".xml")) {
```
[the remark this answers is not in the document]
Comment (jbartolotta-sfdc) [reply src/lib/CustomRulePathManager.ts:249]: This allows users to add XML files that aren't in a jar. The PMD CLI supports this use case. This question in the slack channel has more context https://platformcloud.slack.com/archives/CR89AN6ET/p1592506039128700 The follow on comments in that slack channel are also relevant.

**Comment by jbartolotta-sfdc** [comment]
> I think this all looks pretty good, but just to be clear, does the make it so we can use folders containing PMD rules without needing to wrap them in JARS? Or am I misunderstanding something?

That's correct, it allows users to add an XML file that isn't contained in a Jar. This use case is supported by the PMD CLI and the Java code had partial support for this. See my question here for more info on how the Java code implied it should work with an XML file  https://platformcloud.slack.com/archives/CR89AN6ET/p1592506039128700

**Pull request merged**